### PR TITLE
bug: Prevent roots from being re-created when using React 18

### PR DESCRIPTION
**Pull request by diogobeda**
### Summary

When dynamically adding components to a page by using, for example, Turbo streams and either manually calling `ReactRailsUJS.mountComponents` or passing `ReactRailsUJS.handleMount` to a Turbo event, there's a bug where the react roots are re-created on a node instead of being re-used. That resets the local state of components that were already mounted and may also cause visual glitches with the component flashing.

### Other Information

I made a [reproduction repo](https://github.com/diogobeda/react-rails-18-root-bug) to showcase the bug happening and as a test case for the fix. Feel free to try it out for yourselves.

Here's a gif from before the fix:

CleanShot 2023-08-31 at 10 37 40

And one from after the fix:

CleanShot 2023-08-31 at 10 40 58

Please let me know if there's anything else I should add to this PR 🙏 

### Pull Request checklist
_Remove this line after checking all the items here. If the item is not applicable to the PR, both check it out and wrap it by `~`._

- [x] Add/update test to cover these changes
- [ ] Update documentation
- [x] Update CHANGELOG file

### CODE REVIEW

File: CHANGELOG.md
Comment (justin808) [inline CHANGELOG.md:10]: do not remove this line

File: CHANGELOG.md
Comment (justin808) [inline CHANGELOG.md:12]: Always provide a URL link, like [done for Shakapacker](https://raw.githubusercontent.com/shakacode/shakapacker/master/CHANGELOG.md):

```
- Fixed creation of assets:precompile if it is missing [PR 325](https://github.com/shakacode/shakapacker/pull/325) by [ahangarha](https://github.com/ahangarha).
```

File: react_ujs/index.js
Comment (justin808) [inline react_ujs/index.js:201]: Is it possible to have a test that fails without this change? And passes with this change?

Can you extract some code from
> I made a [reproduction repo](https://github.com/diogobeda/react-rails-18-root-bug) to showcase the bug happening and as a test case for the fix. Feel free to try it out for yourselves.

into https://github.com/reactjs/react-rails/blob/master/test/dummy

and add a test?

**Review by justin808 — CHANGES_REQUESTED** [review]

File: react_ujs/index.js
Comment (diogobeda) [reply react_ujs/index.js:201]: Sounds good! I'll do that in the next few days.

Comment (diogobeda) [reply react_ujs/index.js:201]: Hey @justin808, the dummy project's appraisals are missing some of the dependencies I've used in the reproduction case, such as Turbo and sqlite for creating the Timer records. Can I add those or would you prefer another approach?

Comment (justin808) [reply react_ujs/index.js:201]: @diogobeda feel free to add them.

**Comment by justin808** [comment]
@diogobeda let me know when you added tests.

We need CI to pass.

**Comment by justin808** [comment]
@diogobeda, I'd like to merge. Let me know when you added tests.

We need CI to pass.

**Comment by diogobeda** [comment]
Hey @justin808, sorry! I haven't been having much time to work on this, but I'll try to prioritize this for next week.

**Comment by justin808** [comment]
@diogobeda thanks! Let's get this PR merged!

**Comment by diogobeda** [comment]
@justin808 added a test case and confirmed it failed on `master` with React 18, so we should be good to go if there's nothing else to fix on CI

**Comment by justin808** [comment]
@ahangarha @Judahmeek @diogobeda are the CI errors legit?

**Comment by diogobeda** [comment]
I was trying these out locally and I can see the failures but I'm not sure what's wrong because I can't debug them for some reason. It's hard trying to log things in those tests and I'm not sure what to do next for them, so I could use some help.

**Comment by justin808** [comment]
Seem integration tests fail with Sprockets 3 and 4, but work with shakapacker. 

@diogobeda does that help?

See 

https://github.com/reactjs/react-rails/blob/master/.github/workflows/ruby.yml#L61

https://github.com/reactjs/react-rails/blob/master/gemfiles/sprockets_3.gemfile

**Comment by justin808** [comment]
@diogobeda please [join our slack](https://join.slack.com/t/reactrails/shared_invite/enQtNjY3NTczMjczNzYxLTlmYjdiZmY3MTVlMzU2YWE0OWM0MzNiZDI0MzdkZGFiZTFkYTFkOGVjODBmOWEyYWQ3MzA2NGE1YWJjNmVlMGE) for react-rails and ping @Judahmeek to pair on this. 

I truly appreciate your help!

**Comment by justin808** [comment]
@ahangarha @G-Rath should this be merged once CI passes? Might be the same issue at #1306 for CI.

**Comment by G-Rath** [comment]
The CI failures here look different to the ones in #1306 - it does look like I broke CI though in #1302 which I've opened #1328 to revert; that might unblock this once landed, but either way it should put us in a better state.

**Comment by justin808** [comment]
@diogobeda can you rebase on master?

**Comment by justin808** [comment]
I'd love to get this merged, but we have spec failures and rubocop issues.

File: test/react/rails/realtime_update_test.rb
Comment (G-Rath) [inline test/react/rails/realtime_update_test.rb:12]: This is what's causing the log error:

```suggestion
        #{page.driver.browser.manage.logs.get(:browser).inspect}
```

**Review by G-Rath — CHANGES_REQUESTED** [review]
I'm not actually able to run the full test suite locally due to my setup (it requires a higher version of `selenium-webdrivers` than what's supported by Ruby 2.7), but besides the `logs` error (which I've suggested a fix for) I'm not see anything indicating this isn't just a general issue in your tests since they're now passing on the default branch.

It would be good if someone who can run the full test suite locally could confirm if this errors are reproducable.

I'm also curious to why you're only installing `turbo-rails` for the `shakapacker` gemfile?

**Comment by justin808** [comment]
@diogobeda can you please check test failures?

File: test/react/rails/realtime_update_test.rb
Comment (G-Rath) [reply test/react/rails/realtime_update_test.rb:12]: I did recommend applying it back when we were doing the CI fixing PRs, but we ended up going the other direction.

Overall that change is out of scope for this PR, but I also think it might also not be possible on the current gem versions which is why I've not done a PR myself - I'm happy to review a PR if you want to give it a try, but it'll eventually get done after dropping support for Ruby 2.7 etc if it's not done before then.

**Comment by diogobeda** [comment]
Hey @justin808 @G-Rath, I'm gonna close this PR because it's been too hard to deal with the project's test suite and I can't keep on top of it. Feel free to take my commits/changes and run with it, but I can no longer contribute.

**Comment by justin808** [comment]
Reopen so I don't forget!

**Pull request closed without merging**
